🐛 BUG: Using both `--route` and `--remote` causes WebSocket error

### Which Cloudflare product(s) does this pertain to?

Wrangler core

### What version(s) of the tool(s) are you using?

3.5.0

### What version of Node are you using?

20.5.1

### What operating system are you using?

macOS Ventura 13.5

### Describe the Bug

1. Create `index.js` file:
```js
export default {
	fetch(req) {
		const url = new URL(req.url);
		return new Response(`Your hostname is: ${url.hostname}`);
	}
}
```
2. Run Wrangler Dev with both `--remote` and `--route` flags(with a zone you have access to):
```sh
$ wrangler dev index.js --route "goalastair.com/*" --remote
 ⛅️ wrangler 3.5.0
------------------
⬣ Listening at http://0.0.0.0:8787
- http://127.0.0.1:8787
- http://192.168.32.2:8787
Total Upload: 5.96 KiB / gzip: 1.84 KiB
▲ [WARNING] worker failed to prewarm:  Bad Gateway


Waiting for connection...
╭────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╮
│ [b] open a browser, [d] open Devtools, [l] turn on local mode, [c] clear console, [x] to exit                                                                                                                  │
╰────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:30632
            throw a;
            ^

Error: WebSocket was closed before the connection was established
    at WebSocket2.close (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:108521:11)
    at close (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:127122:14)
    at /Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:127230:7
    at jh (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:29688:15)
    at exports2.unstable_runWithPriority (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:25660:16)
    at Pc (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:26308:16)
    at Qg (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:29651:18)
    at /Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:29640:13
    at V (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:25585:22)
    at Timeout.w [as _onTimeout] (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:25420:13)
Emitted 'error' event on WebSocket instance at:
    at emitErrorAndClose (/Project/Directory/node_modules/.pnpm/registry.npmjs.org+wrangler@3.5.0/node_modules/wrangler/wrangler-dist/cli.js:108968:17)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21)

Node.js v20.5.1
```

### Please provide a link to a minimal reproduction

_No response_

### Please provide any relevant error logs

_No response_

@iksaku I get a very similar error log running `wrangler dev --remote`, but if I run `wrangler dev` and then press `l` to switch from local to remote, the error is significantly different.

This JSON parsing error is the one that I mentioned in an earlier post that is similar to the Wrangler v2 bug at #2581 (at least in that they're trying to JSON.parse an HTML file when trying launch a remote dev server.

Stack trace:
```
✘ [ERROR] Error while creating remote dev session: ParseError: Unexpected token <

      at parseJSON
  (/home/runcloud/wrangler-projects/telcom-processor/node_modules/wrangler/wrangler-dist/cli.js:119580:11)
      at createPreviewSession
  (/home/runcloud/wrangler-projects/telcom-processor/node_modules/wrangler/wrangler-dist/cli.js:150592:51)
      at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
      at async start
  (/home/runcloud/wrangler-projects/telcom-processor/node_modules/wrangler/wrangler-dist/cli.js:150781:9)
```

My HTML file is a cloudflare error page - Error 1031: Improperly configured Workers Preview for dev-telcom.eaevsf.us. You may need to reinitialize.

Interestingly, I've commented out routes in wrangler.toml and I'm getting this error when _switching to_ remote, but not when using the `--remote` flag. 

This is occurring on the sub.domain/* worker while I have a domain.com/* worker manually configured on the dashboard. (This is the live services setup, so I can't test the effect of taking down the domain.com/* route binding.

This is extremely frustrating as workerd is still encountering some issues that are not present when deployed. I'd really appreciate this being looked on more. 

I often get:
`⚠️ [WARNING] worker failed to prewarm:  Bad Gateway`
OR
`✘ [ERROR] Error while creating remote dev session: ParseError: Unexpected token <

      at parseJSON
`

The workaround using `npx wrangler dev --remote --routes "[]"` does help :)

--- a/.changeset/tame-needles-give.md
+++ b/.changeset/tame-needles-give.md
@@ -1,0 +1,5 @@
+---
+"wrangler": patch
+---
+
+fix: Ensures that switching to remote mode during a dev session (from local mode) will correctly use the right zone. Previously, zone detection happened before the dev session was mounted, and so dev sessions started with local mode would have no zone inferred, and would have failed to start, with an ugly error.

--- a/.changeset/tasty-jokes-provide.md
+++ b/.changeset/tasty-jokes-provide.md
@@ -1,0 +1,5 @@
+---
+"wrangler": patch
+---
+
+fix: Ensure that preview sessions created without a zone don't switch the host on which to start the preview from the one returned by the API.

--- a/packages/wrangler/e2e/dev.test.ts
+++ b/packages/wrangler/e2e/dev.test.ts
@@ -734,7 +734,7 @@ describe("writes debug logs to hidden file", () => {
 	});
 });
 
-describe.only("zone selection", () => {
+describe("zone selection", () => {
 	let worker: DevWorker;
 
 	beforeEach(async () => {

--- a/packages/wrangler/e2e/dev.test.ts
+++ b/packages/wrangler/e2e/dev.test.ts
@@ -733,3 +733,122 @@ describe("writes debug logs to hidden file", () => {
 		});
 	});
 });
+
+describe.only("zone selection", () => {
+	let worker: DevWorker;
+
+	beforeEach(async () => {
+		worker = await makeWorker();
+		await worker.seed((workerName) => ({
+			"wrangler.toml": dedent`
+					name = "${workerName}"
+					main = "src/index.ts"
+					compatibility_date = "2023-01-01"
+					compatibility_flags = ["nodejs_compat"]
+			`,
+			"src/index.ts": dedent`
+					export default {
+						fetch(request) {
+							return new Response(request.url)
+						}
+					}`,
+			"package.json": dedent`
+					{
+						"name": "${workerName}",
+						"version": "0.0.0",
+						"private": true
+					}
+					`,
+		}));
+	});
+
+	it("defaults to a workers.dev preview", async () => {
+		await worker.runDevSession("--remote --ip 127.0.0.1", async (session) => {
+			const { text } = await retry(
+				(s) => s.status !== 200,
+				async () => {
+					const r = await fetch(`http://127.0.0.1:${session.port}`);
+					return { text: await r.text(), status: r.status };
+				}
+			);
+			expect(text).toContain(`devprod-testing7928.workers.dev`);
+		});
+	});
+
+	it("respects dev.host setting", async () => {
+		await worker.seed((workerName) => ({
+			"wrangler.toml": dedent`
+					name = "${workerName}"
+					main = "src/index.ts"
+					compatibility_date = "2023-01-01"
+					compatibility_flags = ["nodejs_compat"]
+
+					[dev]
+					host = "wrangler-testing.testing.devprod.cloudflare.dev"
+			`,
+		}));
+		await worker.runDevSession("--remote --ip 127.0.0.1", async (session) => {
+			const { text } = await retry(
+				(s) => s.status !== 200,
+				async () => {
+					const r = await fetch(`http://127.0.0.1:${session.port}`);
+					return { text: await r.text(), status: r.status };
+				}
+			);
+			expect(text).toMatchInlineSnapshot(
+				`"https://wrangler-testing.testing.devprod.cloudflare.dev/"`
+			);
+		});
+	});
+	it("infers host from first route", async () => {
+		await worker.seed((workerName) => ({
+			"wrangler.toml": dedent`
+					name = "${workerName}"
+					main = "src/index.ts"
+					compatibility_date = "2023-01-01"
+					compatibility_flags = ["nodejs_compat"]
+
+					[[routes]]
+					pattern = "wrangler-testing.testing.devprod.cloudflare.dev/*"
+					zone_name = "testing.devprod.cloudflare.dev"
+			`,
+		}));
+		await worker.runDevSession("--remote --ip 127.0.0.1", async (session) => {
+			const { text } = await retry(
+				(s) => s.status !== 200,
+				async () => {
+					const r = await fetch(`http://127.0.0.1:${session.port}`);
+					return { text: await r.text(), status: r.status };
+				}
+			);
+			expect(text).toMatchInlineSnapshot(
+				`"https://wrangler-testing.testing.devprod.cloudflare.dev/"`
+			);
+		});
+	});
+	it("fails with useful error message if host is not routable", async () => {
+		await worker.seed((workerName) => ({
+			"wrangler.toml": dedent`
+					name = "${workerName}"
+					main = "src/index.ts"
+					compatibility_date = "2023-01-01"
+					compatibility_flags = ["nodejs_compat"]
+
+					[[routes]]
+					pattern = "not-a-domain.testing.devprod.cloudflare.dev/*"
+					zone_name = "testing.devprod.cloudflare.dev"
+			`,
+		}));
+		await worker.runDevSession("--remote --ip 127.0.0.1", async (session) => {
+			await setTimeout(5000);
+			expect(session.stderr).toMatchInlineSnapshot(`
+				"[31m✘ [41;31m[[41;97mERROR[41;31m][0m [1mCould not access \`not-a-domain.testing.devprod.cloudflare.dev\`. Make sure the domain is set up to be proxied by Cloudflare.[0m
+
+				  For more details, refer to [4mhttps://developers.cloudflare.com/workers/configuration/routing/routes/#set-up-a-route[0m
+
+
+				"
+			`);
+		});
+	});
+});

--- a/packages/wrangler/e2e/dev.test.ts
+++ b/packages/wrangler/e2e/dev.test.ts
@@ -840,8 +840,13 @@ describe("zone selection", () => {
 			`,
 		}));
 		await worker.runDevSession("--remote --ip 127.0.0.1", async (session) => {
-			await setTimeout(5000);
-			expect(session.stderr).toMatchInlineSnapshot(`
+			const { stderr } = await retry(
+				(s) => !s.stderr.includes("ERROR"),
+				() => {
+					return { stderr: session.stderr };
+				}
+			);
+			expect(stderr).toMatchInlineSnapshot(`
 				"[31m✘ [41;31m[[41;97mERROR[41;31m][0m [1mCould not access \`not-a-domain.testing.devprod.cloudflare.dev\`. Make sure the domain is set up to be proxied by Cloudflare.[0m
 
 				  For more details, refer to [4mhttps://developers.cloudflare.com/workers/configuration/routing/routes/#set-up-a-route[0m

--- a/packages/wrangler/e2e/dev.test.ts
+++ b/packages/wrangler/e2e/dev.test.ts
@@ -846,13 +846,9 @@ describe("zone selection", () => {
 					return { stderr: session.stderr };
 				}
 			);
-			expect(stderr).toMatchInlineSnapshot(`
-				"[31m✘ [41;31m[[41;97mERROR[41;31m][0m [1mCould not access \`not-a-domain.testing.devprod.cloudflare.dev\`. Make sure the domain is set up to be proxied by Cloudflare.[0m
-
-				  For more details, refer to [4mhttps://developers.cloudflare.com/workers/configuration/routing/routes/#set-up-a-route[0m
-
-
-				"
+			expect(normalizeOutput(stderr)).toMatchInlineSnapshot(`
+				"X [ERROR] Could not access \`not-a-domain.testing.devprod.cloudflare.dev\`. Make sure the domain is set up to be proxied by Cloudflare.
+				  For more details, refer to https://developers.cloudflare.com/workers/configuration/routing/routes/#set-up-a-route"
 			`);
 		});
 	});

--- a/packages/wrangler/e2e/helpers/account-id.ts
+++ b/packages/wrangler/e2e/helpers/account-id.ts
@@ -1,1 +1,2 @@
-export const CLOUDFLARE_ACCOUNT_ID = process.env.CLOUDFLARE_ACCOUNT_ID;
+export const CLOUDFLARE_ACCOUNT_ID = process.env
+	.CLOUDFLARE_ACCOUNT_ID as string;

--- a/packages/wrangler/e2e/helpers/account-id.ts
+++ b/packages/wrangler/e2e/helpers/account-id.ts
@@ -1,8 +1,1 @@
-import assert from "node:assert";
-
-assert(
-	process.env.CLOUDFLARE_ACCOUNT_ID,
-	"Please provide a CLOUDFLARE_ACCOUNT_ID as an environment variable"
-);
-
 export const CLOUDFLARE_ACCOUNT_ID = process.env.CLOUDFLARE_ACCOUNT_ID;

--- a/packages/wrangler/e2e/helpers/wrangler-command.ts
+++ b/packages/wrangler/e2e/helpers/wrangler-command.ts
@@ -1,1 +1,1 @@
-export const WRANGLER = process.env.WRANGLER;
+export const WRANGLER = process.env.WRANGLER as string;

--- a/packages/wrangler/e2e/helpers/wrangler-command.ts
+++ b/packages/wrangler/e2e/helpers/wrangler-command.ts
@@ -1,2 +1,1 @@
-export const WRANGLER =
-	process.env.WRANGLER ?? `pnpm --silent dlx wrangler@beta`;
+export const WRANGLER = process.env.WRANGLER;

--- a/packages/wrangler/e2e/validate-environment.ts
+++ b/packages/wrangler/e2e/validate-environment.ts
@@ -1,0 +1,23 @@
+import assert from "node:assert";
+
+assert(
+	process.env.WRANGLER,
+	'You must provide a way to run Wrangler (WRANGLER="pnpm --silent dlx wrangler@beta" will run the latest beta)'
+);
+
+assert(
+	process.env.CLOUDFLARE_ACCOUNT_ID,
+	"You must provide a CLOUDFLARE_ACCOUNT_ID as an environment variable"
+);
+
+assert(
+	process.env.CLOUDFLARE_API_TOKEN,
+	"You must provide a CLOUDFLARE_API_TOKEN as an environment variable"
+);
+
+assert(
+	process.env.CLOUDFLARE_ACCOUNT_ID === "8d783f274e1f82dc46744c297b015a2f",
+	"You must run Wrangler's e2e tests against DevProd Testing"
+);
+
+export const setup = () => {};

--- a/packages/wrangler/e2e/vitest.config.ts
+++ b/packages/wrangler/e2e/vitest.config.ts
@@ -1,0 +1,16 @@
+import path from "path";
+import { defineConfig } from "vitest/config";
+
+export default defineConfig({
+	test: {
+		testTimeout: 240_000,
+		poolOptions: {
+			threads: {
+				singleThread: true,
+			},
+		},
+		retry: 2,
+		include: ["e2e/**/*.test.ts"],
+		globalSetup: path.resolve(__dirname, "./validate-environment.ts"),
+	},
+});

--- a/packages/wrangler/package.json
+++ b/packages/wrangler/package.json
@@ -63,7 +63,7 @@
 		"test": "pnpm run assert-git-version && jest",
 		"test:ci": "pnpm run test --coverage",
 		"test:debug": "pnpm run test --silent=false --verbose=true",
-		"test:e2e": "vitest --test-timeout 240000 --poolOptions.threads.singleThread --dir ./e2e --retry 2 run",
+		"test:e2e": "vitest -c ./e2e/vitest.config.ts",
 		"test:watch": "pnpm run test --testTimeout=50000 --watch",
 		"type:tests": "tsc -p ./src/__tests__/tsconfig.json && tsc -p ./e2e/tsconfig.json"
 	},

--- a/packages/wrangler/src/__tests__/dev.test.tsx
+++ b/packages/wrangler/src/__tests__/dev.test.tsx
@@ -5,6 +5,7 @@ import { rest } from "msw";
 import patchConsole from "patch-console";
 import dedent from "ts-dedent";
 import Dev from "../dev/dev";
+import { getWorkerAccountAndContext } from "../dev/remote";
 import { CI } from "../is-ci";
 import { mockAccountId, mockApiToken } from "./helpers/mock-account-id";
 import { mockConsoleMethods } from "./helpers/mock-console";
@@ -17,6 +18,36 @@ import {
 import { runInTempDir } from "./helpers/run-in-tmp";
 import { runWrangler } from "./helpers/run-wrangler";
 import writeWranglerToml from "./helpers/write-wrangler-toml";
+
+async function expectedHostAndZone(
+	host: string,
+	zone: string
+): Promise<unknown> {
+	const config = (Dev as jest.Mock).mock.calls[0][0];
+	expect(config).toEqual(
+		expect.objectContaining({
+			localUpstream: host,
+		})
+	);
+	await expect(
+		getWorkerAccountAndContext({
+			accountId: "",
+			host: config.host,
+			routes: config.routes,
+		})
+	).resolves.toEqual(
+		expect.objectContaining({
+			workerContext: {
+				host,
+				zone,
+				routes: config.routes,
+			},
+		})
+	);
+
+	(Dev as jest.Mock).mockClear();
+	return config;
+}
 
 describe("wrangler dev", () => {
 	beforeEach(() => {
@@ -213,10 +244,14 @@ describe("wrangler dev", () => {
 				routes: ["http://5.some-host.com/some/path/*"],
 			});
 			await runWrangler("dev --remote");
-			expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
+
+			const devConfig = await expectedHostAndZone(
+				"5.some-host.com",
+				"some-zone-id-5"
+			);
+
+			expect(devConfig).toEqual(
 				expect.objectContaining({
-					host: "5.some-host.com",
-					zone: "some-zone-id-5",
 					routes: ["http://5.some-host.com/some/path/*"],
 				})
 			);
@@ -230,12 +265,8 @@ describe("wrangler dev", () => {
 			fs.writeFileSync("index.js", `export default {};`);
 			mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
 			await runWrangler("dev --remote --host some-host.com");
-			expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
-				expect.objectContaining({
-					host: "some-host.com",
-					zone: "some-zone-id",
-				})
-			);
+
+			await expectedHostAndZone("some-host.com", "some-zone-id");
 		});
 
 		it("should read wrangler.toml's dev.host", async () => {
@@ -258,7 +289,7 @@ describe("wrangler dev", () => {
 			fs.writeFileSync("index.js", `export default {};`);
 			mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
 			await runWrangler("dev --route http://some-host.com/some/path/*");
-			expect((Dev as jest.Mock).mock.calls[0][0].host).toEqual("some-host.com");
+			await expectedHostAndZone("some-host.com", "some-zone-id");
 		});
 
 		it("should read wrangler.toml's routes", async () => {
@@ -272,7 +303,7 @@ describe("wrangler dev", () => {
 			fs.writeFileSync("index.js", `export default {};`);
 			mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
 			await runWrangler("dev");
-			expect((Dev as jest.Mock).mock.calls[0][0].host).toEqual("some-host.com");
+			await expectedHostAndZone("some-host.com", "some-zone-id");
 		});
 
 		it("should read wrangler.toml's environment specific routes", async () => {
@@ -294,7 +325,7 @@ describe("wrangler dev", () => {
 			fs.writeFileSync("index.js", `export default {};`);
 			mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
 			await runWrangler("dev --env staging");
-			expect((Dev as jest.Mock).mock.calls[0][0].host).toEqual("some-host.com");
+			await expectedHostAndZone("some-host.com", "some-zone-id");
 		});
 
 		it("should strip leading `*` from given host when deducing a zone id", async () => {
@@ -305,7 +336,7 @@ describe("wrangler dev", () => {
 			fs.writeFileSync("index.js", `export default {};`);
 			mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
 			await runWrangler("dev");
-			expect((Dev as jest.Mock).mock.calls[0][0].host).toEqual("some-host.com");
+			await expectedHostAndZone("some-host.com", "some-zone-id");
 		});
 
 		it("should strip leading `*.` from given host when deducing a zone id", async () => {
@@ -316,7 +347,7 @@ describe("wrangler dev", () => {
 			fs.writeFileSync("index.js", `export default {};`);
 			mockGetZones("some-host.com", [{ id: "some-zone-id" }]);
 			await runWrangler("dev");
-			expect((Dev as jest.Mock).mock.calls[0][0].host).toEqual("some-host.com");
+			await expectedHostAndZone("some-host.com", "some-zone-id");
 		});
 
 		it("should, when provided, use a configured zone_id", async () => {
@@ -328,12 +359,8 @@ describe("wrangler dev", () => {
 			});
 			fs.writeFileSync("index.js", `export default {};`);
 			await runWrangler("dev --remote");
-			expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
-				expect.objectContaining({
-					host: "some-domain.com",
-					zone: "some-zone-id",
-				})
-			);
+
+			await expectedHostAndZone("some-domain.com", "some-zone-id");
 		});
 
 		it("should, when provided, use a zone_name to get a zone_id", async () => {
@@ -349,13 +376,8 @@ describe("wrangler dev", () => {
 			fs.writeFileSync("index.js", `export default {};`);
 			mockGetZones("some-zone.com", [{ id: "a-zone-id" }]);
 			await runWrangler("dev --remote");
-			expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
-				expect.objectContaining({
-					// note that it uses the provided zone_name as a host too
-					host: "some-zone.com",
-					zone: "a-zone-id",
-				})
-			);
+
+			await expectedHostAndZone("some-zone.com", "a-zone-id");
 		});
 
 		it("should find the host from the given pattern, not zone_name", async () => {
@@ -385,9 +407,11 @@ describe("wrangler dev", () => {
 				],
 			});
 			await fs.promises.writeFile("index.js", `export default {};`);
-			await expect(runWrangler("dev --remote")).rejects.toEqual(
-				new Error("Could not find zone for does-not-exist.com")
-			);
+			await expect(runWrangler("dev --remote")).rejects
+				.toThrowErrorMatchingInlineSnapshot(`
+			"Could not find zone for \`does-not-exist.com\`. Make sure the domain is set up to be proxied by Cloudflare.
+			For more details, refer to https://developers.cloudflare.com/workers/configuration/routing/routes/#set-up-a-route"
+		`);
 		});
 
 		it("should fallback to zone_name when given the pattern */*", async () => {
@@ -464,14 +488,7 @@ describe("wrangler dev", () => {
 
 			await runWrangler("dev --remote --host 111.222.333.some-host.com");
 
-			const devMockCall = (Dev as jest.Mock).mock.calls[0][0];
-
-			expect(devMockCall).toHaveProperty("host", "111.222.333.some-host.com");
-			expect(devMockCall).toHaveProperty(
-				"localUpstream",
-				"111.222.333.some-host.com"
-			);
-			expect(devMockCall).toHaveProperty("zone", "some-zone-id");
+			await expectedHostAndZone("111.222.333.some-host.com", "some-zone-id");
 		});
 
 		it("should, in order, use args.host/config.dev.host/args.routes/(config.route|config.routes)", async () => {
@@ -486,13 +503,8 @@ describe("wrangler dev", () => {
 				routes: ["http://5.some-host.com/some/path/*"],
 			});
 			await runWrangler("dev --remote");
-			expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
-				expect.objectContaining({
-					host: "5.some-host.com",
-					zone: "some-zone-id-5",
-				})
-			);
-			(Dev as jest.Mock).mockClear();
+
+			await expectedHostAndZone("5.some-host.com", "some-zone-id-5");
 
 			// config.route
 			mockGetZones("4.some-host.com", [{ id: "some-zone-id-4" }]);
@@ -501,13 +513,8 @@ describe("wrangler dev", () => {
 				route: "https://4.some-host.com/some/path/*",
 			});
 			await runWrangler("dev --remote");
-			expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
-				expect.objectContaining({
-					host: "4.some-host.com",
-					zone: "some-zone-id-4",
-				})
-			);
-			(Dev as jest.Mock).mockClear();
+
+			await expectedHostAndZone("4.some-host.com", "some-zone-id-4");
 
 			// --routes
 			mockGetZones("3.some-host.com", [{ id: "some-zone-id-3" }]);
@@ -518,13 +525,8 @@ describe("wrangler dev", () => {
 			await runWrangler(
 				"dev --remote --routes http://3.some-host.com/some/path/*"
 			);
-			expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
-				expect.objectContaining({
-					host: "3.some-host.com",
-					zone: "some-zone-id-3",
-				})
-			);
-			(Dev as jest.Mock).mockClear();
+
+			await expectedHostAndZone("3.some-host.com", "some-zone-id-3");
 
 			// config.dev.host
 			mockGetZones("2.some-host.com", [{ id: "some-zone-id-2" }]);
@@ -538,13 +540,7 @@ describe("wrangler dev", () => {
 			await runWrangler(
 				"dev --remote --routes http://3.some-host.com/some/path/*"
 			);
-			expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
-				expect.objectContaining({
-					host: "2.some-host.com",
-					zone: "some-zone-id-2",
-				})
-			);
-			(Dev as jest.Mock).mockClear();
+			await expectedHostAndZone("2.some-host.com", "some-zone-id-2");
 
 			// --host
 			mockGetZones("1.some-host.com", [{ id: "some-zone-id-1" }]);
@@ -558,13 +554,7 @@ describe("wrangler dev", () => {
 			await runWrangler(
 				"dev --remote --routes http://3.some-host.com/some/path/* --host 1.some-host.com"
 			);
-			expect((Dev as jest.Mock).mock.calls[0][0]).toEqual(
-				expect.objectContaining({
-					host: "1.some-host.com",
-					zone: "some-zone-id-1",
-				})
-			);
-			(Dev as jest.Mock).mockClear();
+			await expectedHostAndZone("1.some-host.com", "some-zone-id-1");
 		});
 
 		it("should error if a host can't resolve to a zone", async () => {
@@ -573,11 +563,11 @@ describe("wrangler dev", () => {
 			});
 			fs.writeFileSync("index.js", `export default {};`);
 			mockGetZones("some-host.com", []);
-			await expect(
-				runWrangler("dev --remote --host some-host.com")
-			).rejects.toThrowErrorMatchingInlineSnapshot(
-				`"Could not find zone for some-host.com"`
-			);
+			await expect(runWrangler("dev --remote --host some-host.com")).rejects
+				.toThrowErrorMatchingInlineSnapshot(`
+			"Could not find zone for \`some-host.com\`. Make sure the domain is set up to be proxied by Cloudflare.
+			For more details, refer to https://developers.cloudflare.com/workers/configuration/routing/routes/#set-up-a-route"
+		`);
 		});
 
 		it("should not try to resolve a zone when starting in local mode", async () => {
@@ -612,7 +602,7 @@ describe("wrangler dev", () => {
 			});
 			fs.writeFileSync("index.js", `export default {};`);
 			await runWrangler("dev");
-			expect((Dev as jest.Mock).mock.calls[0][0].host).toEqual(
+			expect((Dev as jest.Mock).mock.calls[0][0].localUpstream).toEqual(
 				"4.some-host.com"
 			);
 		});

--- a/packages/wrangler/src/__tests__/zones.test.ts
+++ b/packages/wrangler/src/__tests__/zones.test.ts
@@ -69,11 +69,11 @@ describe("Zones", () => {
 
 		test("string route (not a zone)", async () => {
 			mockGetZones("wrong.com", []);
-			await expect(
-				getZoneForRoute("wrong.com/*")
-			).rejects.toMatchInlineSnapshot(
-				`[Error: Could not find zone for wrong.com]`
-			);
+			await expect(getZoneForRoute("wrong.com/*")).rejects
+				.toMatchInlineSnapshot(`
+			[Error: Could not find zone for \`wrong.com\`. Make sure the domain is set up to be proxied by Cloudflare.
+			For more details, refer to https://developers.cloudflare.com/workers/configuration/routing/routes/#set-up-a-route]
+		`);
 		});
 		test("zone_id route", async () => {
 			// example-id and other-id intentionally different to show that the API is not called

--- a/packages/wrangler/src/dev.tsx
+++ b/packages/wrangler/src/dev.tsx
@@ -404,7 +404,6 @@ export async function startDev(args: StartDevOptions) {
 			legacyNodeCompat,
 			nodejsCompat,
 			upstreamProtocol,
-			zoneId,
 			host,
 			routes,
 			getLocalPort,
@@ -439,7 +438,6 @@ export async function startDev(args: StartDevOptions) {
 					findAdditionalModules={configParam.find_additional_modules}
 					entry={entry}
 					env={args.env}
-					zone={zoneId}
 					host={host}
 					routes={routes}
 					processEntrypoint={processEntrypoint}
@@ -536,7 +534,6 @@ export async function startApiDev(args: StartDevOptions) {
 		legacyNodeCompat,
 		nodejsCompat,
 		upstreamProtocol,
-		zoneId,
 		host,
 		routes,
 		getLocalPort,
@@ -571,7 +568,6 @@ export async function startApiDev(args: StartDevOptions) {
 			findAdditionalModules: configParam.find_additional_modules,
 			entry: entry,
 			env: args.env,
-			zone: zoneId,
 			host: host,
 			routes: routes,
 			processEntrypoint,
@@ -715,7 +711,7 @@ async function validateDevServerSettings(
 		"dev"
 	);
 
-	const { zoneId, host, routes } = await getHostAndRoutes(args, config);
+	const { host, routes } = await getHostAndRoutes(args, config);
 	const initialIp = args.ip || config.dev.ip;
 	const initialIpListenCheck = initialIp === "*" ? "0.0.0.0" : initialIp;
 	const getLocalPort = memoizeGetPort(DEFAULT_LOCAL_PORT, initialIpListenCheck);
@@ -815,7 +811,6 @@ async function validateDevServerSettings(
 		getLocalPort,
 		getInspectorPort,
 		getRuntimeInspectorPort,
-		zoneId,
 		host,
 		routes,
 		cliDefines,

--- a/packages/wrangler/src/dev.tsx
+++ b/packages/wrangler/src/dev.tsx
@@ -16,7 +16,7 @@ import * as metrics from "./metrics";
 import { getAssetPaths, getSiteAssetPaths } from "./sites";
 import { getAccountFromCache, loginOrRefreshIfRequired } from "./user";
 import { collectKeyValues } from "./utils/collectKeyValues";
-import { getHostFromRoute } from "./zones";
+import { getHostFromRoute, getZoneIdForPreview } from "./zones";
 import {
 	DEFAULT_INSPECTOR_PORT,
 	DEFAULT_LOCAL_PORT,
@@ -457,7 +457,7 @@ export async function startDev(args: StartDevOptions) {
 					localProtocol={args.localProtocol || configParam.dev.local_protocol}
 					httpsKeyPath={args.httpsKeyPath}
 					httpsCertPath={args.httpsCertPath}
-					localUpstream={args.localUpstream ?? host}
+					localUpstream={args.localUpstream ?? host ?? getInferredHost(routes)}
 					localPersistencePath={localPersistencePath}
 					liveReload={args.liveReload || false}
 					accountId={
@@ -587,7 +587,7 @@ export async function startApiDev(args: StartDevOptions) {
 			localProtocol: args.localProtocol ?? configParam.dev.local_protocol,
 			httpsKeyPath: args.httpsKeyPath,
 			httpsCertPath: args.httpsCertPath,
-			localUpstream: args.localUpstream ?? host,
+			localUpstream: args.localUpstream ?? host ?? getInferredHost(routes),
 			localPersistencePath,
 			liveReload: args.liveReload ?? false,
 			accountId:
@@ -673,32 +673,34 @@ function maskVars(bindings: CfWorkerInit["bindings"], configParam: Config) {
 async function getHostAndRoutes(args: StartDevOptions, config: Config) {
 	// TODO: if worker_dev = false and no routes, then error (only for dev)
 	// Compute zone info from the `host` and `route` args and config;
-	let host = args.host || config.dev.host;
+	const host = args.host || config.dev.host;
 	const routes: Route[] | undefined =
 		args.routes || (config.route && [config.route]) || config.routes;
 
-	if (!host) {
-		if (routes) {
-			const firstRoute = routes[0];
-			host = getHostFromRoute(firstRoute);
+	return { host, routes };
+}
 
-			// TODO(consider): do we need really need to do this? I've added the condition to throw to match the previous implicit behaviour of `new URL()` throwing upon invalid URLs, but could we just continue here without an inferred host?
-			if (host === undefined) {
-				throw new UserError(
-					`Cannot infer host from first route: ${JSON.stringify(
-						firstRoute
-					)}.\nYou can explicitly set the \`dev.host\` configuration in your wrangler.toml file, for example:
+export function getInferredHost(routes: Route[] | undefined) {
+	if (routes) {
+		const firstRoute = routes[0];
+		const host = getHostFromRoute(firstRoute);
+
+		// TODO(consider): do we need really need to do this? I've added the condition to throw to match the previous implicit behaviour of `new URL()` throwing upon invalid URLs, but could we just continue here without an inferred host?
+		if (host === undefined) {
+			throw new UserError(
+				`Cannot infer host from first route: ${JSON.stringify(
+					firstRoute
+				)}.\nYou can explicitly set the \`dev.host\` configuration in your wrangler.toml file, for example:
 
 	\`\`\`
 	[dev]
 	host = "example.com"
 	\`\`\`
 `
-				);
-			}
+			);
 		}
+		return host;
 	}
-	return { host, routes };
 }
 
 async function validateDevServerSettings(
@@ -712,6 +714,14 @@ async function validateDevServerSettings(
 	);
 
 	const { host, routes } = await getHostAndRoutes(args, config);
+
+	// TODO: Remove this hack
+	// This function throws if the zone ID can't be found given the provided host and routes
+	// However, it's called as part of initialising a preview session, which is nested deep within
+	// React/Ink and useEffect()s, which swallow the error and turn it into a logw. Because it's a non-recoverable user error,
+	// we want it to exit the Wrangler process early to allow the user to fix it. Calling it here forces
+	// the error to be thrown where it will correctly exit the Wrangler process
+	if (args.remote) await getZoneIdForPreview(host, routes);
 	const initialIp = args.ip || config.dev.ip;
 	const initialIpListenCheck = initialIp === "*" ? "0.0.0.0" : initialIp;
 	const getLocalPort = memoizeGetPort(DEFAULT_LOCAL_PORT, initialIpListenCheck);

--- a/packages/wrangler/src/dev.tsx
+++ b/packages/wrangler/src/dev.tsx
@@ -16,7 +16,7 @@ import * as metrics from "./metrics";
 import { getAssetPaths, getSiteAssetPaths } from "./sites";
 import { getAccountFromCache, loginOrRefreshIfRequired } from "./user";
 import { collectKeyValues } from "./utils/collectKeyValues";
-import { getHostFromRoute, getZoneForRoute, getZoneIdFromHost } from "./zones";
+import { getHostFromRoute } from "./zones";
 import {
 	DEFAULT_INSPECTOR_PORT,
 	DEFAULT_LOCAL_PORT,
@@ -674,27 +674,14 @@ function maskVars(bindings: CfWorkerInit["bindings"], configParam: Config) {
 	return maskedVars;
 }
 
-async function getZoneIdHostAndRoutes(args: StartDevOptions, config: Config) {
+async function getHostAndRoutes(args: StartDevOptions, config: Config) {
 	// TODO: if worker_dev = false and no routes, then error (only for dev)
 	// Compute zone info from the `host` and `route` args and config;
 	let host = args.host || config.dev.host;
-	let zoneId: string | undefined;
 	const routes: Route[] | undefined =
 		args.routes || (config.route && [config.route]) || config.routes;
 
-	if (args.remote) {
-		if (host) {
-			zoneId = await getZoneIdFromHost(host);
-		}
-		if (!zoneId && routes) {
-			const firstRoute = routes[0];
-			const zone = await getZoneForRoute(firstRoute);
-			if (zone) {
-				zoneId = zone.id;
-				host = zone.host;
-			}
-		}
-	} else if (!host) {
+	if (!host) {
 		if (routes) {
 			const firstRoute = routes[0];
 			host = getHostFromRoute(firstRoute);
@@ -715,7 +702,7 @@ async function getZoneIdHostAndRoutes(args: StartDevOptions, config: Config) {
 			}
 		}
 	}
-	return { host, routes, zoneId };
+	return { host, routes };
 }
 
 async function validateDevServerSettings(
@@ -728,7 +715,7 @@ async function validateDevServerSettings(
 		"dev"
 	);
 
-	const { zoneId, host, routes } = await getZoneIdHostAndRoutes(args, config);
+	const { zoneId, host, routes } = await getHostAndRoutes(args, config);
 	const initialIp = args.ip || config.dev.ip;
 	const initialIpListenCheck = initialIp === "*" ? "0.0.0.0" : initialIp;
 	const getLocalPort = memoizeGetPort(DEFAULT_LOCAL_PORT, initialIpListenCheck);

--- a/packages/wrangler/src/dev/create-worker-preview.ts
+++ b/packages/wrangler/src/dev/create-worker-preview.ts
@@ -132,10 +132,16 @@ function randomId(): string {
 
 // URLs are often relative to the zone. Sometimes the base zone
 // will be grey-clouded, and so the host must be swapped out for
-// the worker route host, which is more likely to be orange-clouded
-function switchHost(originalUrl: string, host?: string): URL {
+// the worker route host, which is more likely to be orange-clouded.
+// However, this switching should only happen if we're running a zone preview
+// rather than a workers.dev preview
+function switchHost(
+	originalUrl: string,
+	host: string | undefined,
+	zonePreview: boolean
+): URL {
 	const url = new URL(originalUrl);
-	url.hostname = host ?? url.hostname;
+	url.hostname = zonePreview ? host ?? url.hostname : url.hostname;
 	return url;
 }
 /**
@@ -158,7 +164,11 @@ export async function createPreviewSession(
 		abortSignal
 	);
 
-	const switchedExchangeUrl = switchHost(exchange_url, ctx.host).toString();
+	const switchedExchangeUrl = switchHost(
+		exchange_url,
+		ctx.host,
+		!!ctx.zone
+	).toString();
 
 	logger.debugWithSanitization(
 		"-- START EXCHANGE API REQUEST:",
@@ -192,8 +202,8 @@ export async function createPreviewSession(
 		id: randomId(),
 		value: token,
 		host: ctx.host ?? inspector.host,
-		inspectorUrl: switchHost(inspector.href, ctx.host),
-		prewarmUrl: switchHost(prewarm, ctx.host),
+		inspectorUrl: switchHost(inspector.href, ctx.host, !!ctx.zone),
+		prewarmUrl: switchHost(prewarm, ctx.host, !!ctx.zone),
 	};
 }
 

--- a/packages/wrangler/src/dev/create-worker-preview.ts
+++ b/packages/wrangler/src/dev/create-worker-preview.ts
@@ -2,8 +2,9 @@ import { URL } from "node:url";
 import { fetch } from "undici";
 import { fetchResult } from "../cfetch";
 import { createWorkerUploadForm } from "../deployment-bundle/create-worker-upload-form";
+import { UserError } from "../errors";
 import { logger } from "../logger";
-import { parseJSON } from "../parse";
+import { ParseError, parseJSON } from "../parse";
 import { getAccessToken } from "../user/access";
 import type {
 	CfWorkerContext,
@@ -189,22 +190,35 @@ export async function createPreviewSession(
 	logger.debugWithSanitization("RESPONSE:", bodyText);
 
 	logger.debug("-- END EXCHANGE API RESPONSE");
+	try {
+		const { inspector_websocket, prewarm, token } = parseJSON<{
+			inspector_websocket: string;
+			token: string;
+			prewarm: string;
+		}>(bodyText);
+		const inspector = new URL(inspector_websocket);
+		inspector.searchParams.append("cf_workers_preview_token", token);
 
-	const { inspector_websocket, prewarm, token } = parseJSON<{
-		inspector_websocket: string;
-		token: string;
-		prewarm: string;
-	}>(bodyText);
-	const inspector = new URL(inspector_websocket);
-	inspector.searchParams.append("cf_workers_preview_token", token);
-
-	return {
-		id: randomId(),
-		value: token,
-		host: ctx.host ?? inspector.host,
-		inspectorUrl: switchHost(inspector.href, ctx.host, !!ctx.zone),
-		prewarmUrl: switchHost(prewarm, ctx.host, !!ctx.zone),
-	};
+		return {
+			id: randomId(),
+			value: token,
+			host: ctx.host ?? inspector.host,
+			inspectorUrl: switchHost(inspector.href, ctx.host, !!ctx.zone),
+			prewarmUrl: switchHost(prewarm, ctx.host, !!ctx.zone),
+		};
+	} catch (e) {
+		if (!(e instanceof ParseError)) {
+			throw e;
+		} else {
+			throw new UserError(
+				`Could not create remote preview session on ${
+					ctx.zone
+						? ` host \`${ctx.host}\` on zone \`${ctx.zone}\``
+						: `your account`
+				}.`
+			);
+		}
+	}
 }
 
 /**

--- a/packages/wrangler/src/dev/dev.tsx
+++ b/packages/wrangler/src/dev/dev.tsx
@@ -165,7 +165,6 @@ export type DevProps = {
 	build: Config["build"];
 	env: string | undefined;
 	legacyEnv: boolean;
-	zone: string | undefined;
 	host: string | undefined;
 	routes: Route[] | undefined;
 	inspect: boolean;
@@ -496,7 +495,6 @@ function DevSession(props: DevSessionProps) {
 			usageModel={props.usageModel}
 			env={props.env}
 			legacyEnv={props.legacyEnv}
-			zone={props.zone}
 			host={props.host}
 			routes={props.routes}
 			onReady={announceAndOnReady}

--- a/packages/wrangler/src/dev/remote.tsx
+++ b/packages/wrangler/src/dev/remote.tsx
@@ -59,7 +59,6 @@ interface RemoteProps {
 	usageModel: "bundled" | "unbound" | undefined;
 	env: string | undefined;
 	legacyEnv: boolean | undefined;
-	zone: string | undefined;
 	host: string | undefined;
 	routes: Route[] | undefined;
 	onReady?:
@@ -88,7 +87,6 @@ export function Remote(props: RemoteProps) {
 		usageModel: props.usageModel,
 		env: props.env,
 		legacyEnv: props.legacyEnv,
-		zone: props.zone,
 		host: props.host,
 		routes: props.routes,
 		onReady: props.onReady,
@@ -154,7 +152,6 @@ interface RemoteWorkerProps {
 	usageModel: "bundled" | "unbound" | undefined;
 	env: string | undefined;
 	legacyEnv: boolean | undefined;
-	zone: string | undefined;
 	host: string | undefined;
 	routes: Route[] | undefined;
 	onReady:
@@ -187,7 +184,6 @@ export function useWorker(
 				accountId: props.accountId,
 				env: props.env,
 				legacyEnv: props.legacyEnv,
-				zone: props.zone,
 				host: props.host,
 				routes: props.routes,
 				sendMetrics: props.sendMetrics,
@@ -230,7 +226,6 @@ export function useWorker(
 		props.host,
 		props.legacyEnv,
 		props.routes,
-		props.zone,
 		props.sendMetrics,
 		restartCounter,
 	]);
@@ -275,7 +270,6 @@ export function useWorker(
 				accountId: props.accountId,
 				env: props.env,
 				legacyEnv: props.legacyEnv,
-				zone: props.zone,
 				host: props.host,
 				routes: props.routes,
 				sendMetrics: props.sendMetrics,
@@ -376,7 +370,6 @@ export function useWorker(
 		props.modules,
 		props.env,
 		props.legacyEnv,
-		props.zone,
 		props.host,
 		props.routes,
 		session,
@@ -483,7 +476,6 @@ export async function getRemotePreviewToken(props: RemoteProps) {
 			accountId: props.accountId,
 			env: props.env,
 			legacyEnv: props.legacyEnv,
-			zone: props.zone,
 			host: props.host,
 			routes: props.routes,
 			sendMetrics: props.sendMetrics,

--- a/packages/wrangler/src/dev/remote.tsx
+++ b/packages/wrangler/src/dev/remote.tsx
@@ -16,6 +16,7 @@ import {
 	saveAccountToCache,
 } from "../user";
 import { getAccessToken } from "../user/access";
+import { getZoneIdForPreview } from "../zones";
 import {
 	createPreviewSession,
 	createWorkerPreview,
@@ -180,14 +181,16 @@ export function useWorker(
 			if (props.accountId === undefined) {
 				return;
 			}
-			const { workerAccount, workerContext } = getWorkerAccountAndContext({
-				accountId: props.accountId,
-				env: props.env,
-				legacyEnv: props.legacyEnv,
-				host: props.host,
-				routes: props.routes,
-				sendMetrics: props.sendMetrics,
-			});
+			const { workerAccount, workerContext } = await getWorkerAccountAndContext(
+				{
+					accountId: props.accountId,
+					env: props.env,
+					legacyEnv: props.legacyEnv,
+					host: props.host,
+					routes: props.routes,
+					sendMetrics: props.sendMetrics,
+				}
+			);
 
 			setSession(
 				await createPreviewSession(
@@ -209,6 +212,8 @@ export function useWorker(
 					"You can either enable local mode by pressing l, or register a workers.dev subdomain here:";
 				const onboardingLink = `https://dash.cloudflare.com/${props.accountId}/workers/onboarding`;
 				logger.error(`${errorMessage}\n${solutionMessage}\n${onboardingLink}`);
+			} else if (err instanceof UserError) {
+				logger.error(err.message);
 			}
 			// we want to log the error, but not end the process
 			// since it could recover after the developer fixes whatever's wrong
@@ -266,14 +271,16 @@ export function useWorker(
 				usageModel: props.usageModel,
 			});
 
-			const { workerAccount, workerContext } = getWorkerAccountAndContext({
-				accountId: props.accountId,
-				env: props.env,
-				legacyEnv: props.legacyEnv,
-				host: props.host,
-				routes: props.routes,
-				sendMetrics: props.sendMetrics,
-			});
+			const { workerAccount, workerContext } = await getWorkerAccountAndContext(
+				{
+					accountId: props.accountId,
+					env: props.env,
+					legacyEnv: props.legacyEnv,
+					host: props.host,
+					routes: props.routes,
+					sendMetrics: props.sendMetrics,
+				}
+			);
 
 			const workerPreviewToken = await createWorkerPreview(
 				init,
@@ -472,7 +479,7 @@ export async function getRemotePreviewToken(props: RemoteProps) {
 			throw error;
 		}
 		const abortController = new AbortController();
-		const { workerAccount, workerContext } = getWorkerAccountAndContext({
+		const { workerAccount, workerContext } = await getWorkerAccountAndContext({
 			accountId: props.accountId,
 			env: props.env,
 			legacyEnv: props.legacyEnv,
@@ -621,24 +628,26 @@ async function createRemoteWorkerInit(props: {
 	return init;
 }
 
-function getWorkerAccountAndContext(props: {
+async function getWorkerAccountAndContext(props: {
 	accountId: string;
 	env?: string;
 	legacyEnv?: boolean;
-	zone?: string;
 	host?: string;
 	routes: Route[] | undefined;
 	sendMetrics?: boolean;
-}): { workerAccount: CfAccount; workerContext: CfWorkerContext } {
+}): Promise<{ workerAccount: CfAccount; workerContext: CfWorkerContext }> {
 	const workerAccount: CfAccount = {
 		accountId: props.accountId,
 		apiToken: requireApiToken(),
 	};
 
+	// What zone should the realish preview for this Worker run on?
+	const zoneId = await getZoneIdForPreview(props.host, props.routes);
+
 	const workerContext: CfWorkerContext = {
 		env: props.env,
 		legacyEnv: props.legacyEnv,
-		zone: props.zone,
+		zone: zoneId,
 		host: props.host,
 		routes: props.routes,
 		sendMetrics: props.sendMetrics,

--- a/packages/wrangler/src/dev/remote.tsx
+++ b/packages/wrangler/src/dev/remote.tsx
@@ -7,6 +7,7 @@ import { helpIfErrorIsSizeOrScriptStartup } from "../deploy/deploy";
 import { printBundleSize } from "../deployment-bundle/bundle-reporter";
 import { getBundleType } from "../deployment-bundle/bundle-type";
 import { withSourceURLs } from "../deployment-bundle/source-url";
+import { getInferredHost } from "../dev";
 import { UserError } from "../errors";
 import { logger } from "../logger";
 import { syncAssets } from "../sites";
@@ -333,7 +334,7 @@ export function useWorker(
 					pathname: workerPreviewToken.inspectorUrl.pathname,
 				},
 				userWorkerInnerUrlOverrides: {
-					hostname: props.host,
+					hostname: props.host ?? getInferredHost(props.routes),
 					port: props.port.toString(),
 				},
 				headers: {
@@ -344,7 +345,11 @@ export function useWorker(
 				proxyLogsToController: true,
 			};
 
-			onReady?.(props.host || "localhost", props.port, proxyData);
+			onReady?.(
+				props.host ?? getInferredHost(props.routes) ?? "localhost",
+				props.port,
+				proxyData
+			);
 		}
 		start().catch((err) => {
 			// we want to log the error, but not end the process
@@ -634,7 +639,7 @@ async function createRemoteWorkerInit(props: {
 	return init;
 }
 
-async function getWorkerAccountAndContext(props: {
+export async function getWorkerAccountAndContext(props: {
 	accountId: string;
 	env?: string;
 	legacyEnv?: boolean;
@@ -654,7 +659,7 @@ async function getWorkerAccountAndContext(props: {
 		env: props.env,
 		legacyEnv: props.legacyEnv,
 		zone: zoneId,
-		host: props.host,
+		host: props.host ?? getInferredHost(props.routes),
 		routes: props.routes,
 		sendMetrics: props.sendMetrics,
 	};

--- a/packages/wrangler/src/dev/remote.tsx
+++ b/packages/wrangler/src/dev/remote.tsx
@@ -212,6 +212,12 @@ export function useWorker(
 					"You can either enable local mode by pressing l, or register a workers.dev subdomain here:";
 				const onboardingLink = `https://dash.cloudflare.com/${props.accountId}/workers/onboarding`;
 				logger.error(`${errorMessage}\n${solutionMessage}\n${onboardingLink}`);
+			} else if (
+				(err.cause as { code: string; hostname: string })?.code === "ENOTFOUND"
+			) {
+				logger.error(
+					`Could not access \`${err.cause.hostname}\`. Make sure the domain is set up to be proxied by Cloudflare.\nFor more details, refer to https://developers.cloudflare.com/workers/configuration/routing/routes/#set-up-a-route`
+				);
 			} else if (err instanceof UserError) {
 				logger.error(err.message);
 			}

--- a/packages/wrangler/src/dev/start-server.ts
+++ b/packages/wrangler/src/dev/start-server.ts
@@ -236,7 +236,6 @@ export async function startDevServer(
 			usageModel: props.usageModel,
 			env: props.env,
 			legacyEnv: props.legacyEnv,
-			zone: props.zone,
 			host: props.host,
 			routes: props.routes,
 			onReady: async (ip, port, proxyData) => {

--- a/packages/wrangler/src/dev/validate-dev-props.ts
+++ b/packages/wrangler/src/dev/validate-dev-props.ts
@@ -1,7 +1,7 @@
 import { UserError } from "../errors";
 import type { DevProps } from "./dev";
 
-export function validateDevProps(props: DevProps) {
+export function validateDevProps(props: Omit<DevProps, "host">) {
 	if (
 		!props.isWorkersSite &&
 		props.assetPaths &&

--- a/packages/wrangler/src/zones.ts
+++ b/packages/wrangler/src/zones.ts
@@ -132,7 +132,9 @@ export async function getZoneIdFromHost(host: string): Promise<string> {
 		hostPieces.shift();
 	}
 
-	throw new UserError(`Could not find zone for ${host}`);
+	throw new UserError(
+		`Could not find zone for \`${host}\`. Make sure the domain is set up to be proxied by Cloudflare.\nFor more details, refer to https://developers.cloudflare.com/workers/configuration/routing/routes/#set-up-a-route`
+	);
 }
 
 /**

--- a/packages/wrangler/src/zones.ts
+++ b/packages/wrangler/src/zones.ts
@@ -93,7 +93,23 @@ export function getHostFromUrl(urlLike: string): string | undefined {
 		return undefined;
 	}
 }
-
+export async function getZoneIdForPreview(
+	host: string | undefined,
+	routes: Route[] | undefined
+) {
+	let zoneId: string | undefined;
+	if (host) {
+		zoneId = await getZoneIdFromHost(host);
+	}
+	if (!zoneId && routes) {
+		const firstRoute = routes[0];
+		const zone = await getZoneForRoute(firstRoute);
+		if (zone) {
+			zoneId = zone.id;
+		}
+	}
+	return zoneId;
+}
 /**
  * Given something that resembles a host, try to infer a zone id from it.
  *

